lib/spack/spack/spec.py

```diff
@@ -4143,12 +4140,13 @@ def cformat(self, *args, **kwargs):
         kwargs.setdefault('color', None)
         return self.format(*args, **kwargs)
 
-    def dep_string(self):
-        return ''.join(" ^" + dep.format() for dep in self.sorted_deps())
-
     def __str__(self):
-        ret = self.format() + self.dep_string()
-        return ret.strip()
+        spec_str = self.format()
+        dependencies = [x for x in self.traverse(root=False)]
+        dependencies.sort(key=lambda x: x.name)
+        for current_dependency in dependencies:
+            spec_str += " ^" + current_dependency.format()
```

Comment:
Try to avoid using `+=` with strings.  This is faster and uses `join()` which is implemented in native code:

```suggestion
        dependencies = [self] + sorted(
            x for x in self.traverse(root=False),
            key=lambda x: x.name
        )
        spec_str = " ^".join(d.format() for d in dependencies)
```

Reply:
Why not just
```python
        dependencies = [self] + sorted(self.traverse(root=False),
                                       key=lambda x: x.name)
        spec_str = " ^".join(d.format() for d in dependencies)
```
Isn't using `y` directly better than `x for x in y`? `Spec.traverse` is a generator itself I believe.

Reply:
Yep even better!

Reply:
Should be addressed by 8df0dc8